File: src/datatable/exceptions.py
#!/usr/bin/env python
#-------------------------------------------------------------------------------
# Copyright 2020 H2O.ai
#
# Permission is hereby granted, free of charge, to any person obtaining a
# copy of this software and associated documentation files (the "Software"),
# to deal in the Software without restriction, including without limitation
# the rights to use, copy, modify, merge, publish, distribute, sublicense,
# and/or sell copies of the Software, and to permit persons to whom the
# Software is furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
# IN THE SOFTWARE.
#-------------------------------------------------------------------------------
import builtins
import sys
import traceback
import warnings
from .lib._datatable import apply_color



class DtException(Exception):
    def __init__(self, message):
        assert isinstance(message, str)
        self.msg = message

    def __str__(self):
        return "".join(_split_backtick_string(self.msg))

    def __repr__(self):
        return self.__class__.__name__ + '(' + repr(str(self)) + ')'


class ImportError(DtException, builtins.ImportError): pass
class IndexError(DtException, builtins.IndexError): pass
class InvalidOperationError(DtException): pass
class IOError(DtException, builtins.IOError): pass
class KeyError(DtException, builtins.KeyError): pass
class MemoryError(DtException, builtins.MemoryError): pass
class NotImplementedError(DtException, builtins.NotImplementedError): pass
class OverflowError(DtException, builtins.OverflowError): pass
class TypeError(DtException, builtins.TypeError): pass
class ValueError(DtException, builtins.ValueError): pass


class DatatableWarning(DtException, UserWarning): pass
class IOWarning(DatatableWarning): pass



#-------------------------------------------------------------------------------
# Custom exception handling
#-------------------------------------------------------------------------------

def _handle_dt_exception(exc_class, exc, tb):
    if not isinstance(exc, DtException):
        return _previous_except_hook(exc_class, exc, tb)

    def dim(line):
        return apply_color("dim", line)

    out = ""

    tbframes = traceback.extract_tb(tb)
    col1 = []
    site_packages_dir = None
    last_code_line = None
    for i, frame in enumerate(tbframes):
        ffile = frame.filename
        if ffile == "<stdin>" and frame.name == "<module>":
            col1.append(None)
            continue
        if "site-packages" in ffile:
            if site_packages_dir is None:
                site_packages_dir = ffile[:ffile.index("site-packages")+14]
            if ffile.startswith(site_packages_dir):
                ffile = "$PY/" + ffile[len(site_packages_dir):]
        else:
            last_code_line = i
        line = "%s:%d in %s" % (ffile, frame.lineno, frame.name)
        col1.append(line)

    if col1[-1]:
        out += dim("Traceback (most recent call last):\n")
        col1len = max(len(line) for line in col1 if line) + 2
        prev_line = None
        prev_count = 0
        for i, line1 in enumerate(col1):
            if not line1: continue
            if line1 == prev_line:
                prev_count += 1
                if prev_count >= 3:
                    continue
            else:
                if prev_count >= 3:
                    out += dim("  ... [previous line repeated %d more times]\n"
                               % (prev_count - 3))
                prev_count = 0
                prev_line = line1
            lineout = ("  " + line1 + " "*(col1len - len(line1)) +
                       tbframes[i].line + "\n")
            if i == last_code_line:
                out += lineout
            else:
                out += dim(lineout)
        if prev_count >= 3:
            out += dim("  ... [previous line repeated %d more times]\n"
                       % (prev_count - 3))
        if site_packages_dir:
            out += dim("  (with $PY = %s)\n" % site_packages_dir)
        out += "\n"

    # Lastly, print the exception name & message
    # Also, make sure that any component surrounded with backticks (`like
    # this`) is emphasized
    out += apply_color("red", exc_class.__name__ + ": ")
    # for i, part in enumerate(re.split(r"`(.*?)`", exc.msg)):
    for i, part in enumerate(_split_backtick_string(exc.msg)):
        out += apply_color("bold" if i % 2 else "bright_red", part)
    print(out, file=sys.stderr)


_previous_except_hook = sys.excepthook
sys.excepthook = _handle_dt_exception


def _split_backtick_string(string):
    r"""
    Helper function for processing an exception message with
    backticks. This function will split the string at the
    backtick characters, while also taking care of unescaping any
    escaped special symbols. For example:

        "abc" -> ["abc"]
        "a`b`c" -> ["a", "b", "c"]
        "`abc`" -> ["", "abc"]
        "`\`a\\bc\``" -> ["", "`a\bc`"]
    """
    out = []
    part = ""
    escape_next = False
    for ch in string:
        if escape_next:
            part += ch
            escape_next = False
        elif ch == '\\':
            escape_next = True
        elif ch == '`':
            out.append(part)
            part = ""
        else:
            part += ch
    if part:
        out.append(part)
    return out



#-------------------------------------------------------------------------------
# Custom warning handling
#-------------------------------------------------------------------------------

def _handle_dt_warning(message, category, filename, lineno, file=None,
                       line=None):
    if not issubclass(category, DatatableWarning):
        return _previous_warnings_hook(message, category, filename, lineno,
                                       file, line)
    print(apply_color("yellow", category.__name__ + ": ") +
          apply_color("grey", str(message)),
          file=sys.stderr)


_previous_warnings_hook = warnings.showwarning
warnings.showwarning = _handle_dt_warning

File: src/core/utils/exceptions.cc
//------------------------------------------------------------------------------
// Copyright 2018-2021 H2O.ai
//
// Permission is hereby granted, free of charge, to any person obtaining a
// copy of this software and associated documentation files (the "Software"),
// to deal in the Software without restriction, including without limitation
// the rights to use, copy, modify, merge, publish, distribute, sublicense,
// and/or sell copies of the Software, and to permit persons to whom the
// Software is furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
// IN THE SOFTWARE.
//------------------------------------------------------------------------------
#include <algorithm>
#include <iostream>
#include <errno.h>
#include <string.h>
#include "cstring.h"
#include "ltype.h"
#include "parallel/api.h"
#include "progress/progress_manager.h"
#include "python/obj.h"
#include "python/python.h"
#include "python/string.h"
#include "python/tuple.h"
#include "stype.h"
#include "utils/assert.h"
#include "utils/exceptions.h"


// Singleton, used to write the current "errno" into the stream
CErrno Errno;

static PyObject* DtExc_ImportError           = PyExc_Exception;
static PyObject* DtExc_IndexError            = PyExc_Exception;
static PyObject* DtExc_InvalidOperationError = PyExc_Exception;
static PyObject* DtExc_IOError               = PyExc_Exception;
static PyObject* DtExc_KeyError              = PyExc_Exception;
static PyObject* DtExc_MemoryError           = PyExc_Exception;
static PyObject* DtExc_NotImplementedError   = PyExc_Exception;
static PyObject* DtExc_OverflowError         = PyExc_Exception;
static PyObject* DtExc_TypeError             = PyExc_Exception;
static PyObject* DtExc_ValueError            = PyExc_Exception;
static PyObject* DtWrn_DatatableWarning      = PyExc_Exception;
static PyObject* DtWrn_IOWarning             = PyExc_Exception;

void init_exceptions() {
  auto dx = py::oobj::import("datatable", "exceptions");
  DtExc_ImportError           = dx.get_attr("ImportError").release();
  DtExc_IndexError            = dx.get_attr("IndexError").release();
  DtExc_InvalidOperationError = dx.get_attr("InvalidOperationError").release();
  DtExc_IOError               = dx.get_attr("IOError").release();
  DtExc_KeyError              = dx.get_attr("KeyError").release();
  DtExc_MemoryError           = dx.get_attr("MemoryError").release();
  DtExc_NotImplementedError   = dx.get_attr("NotImplementedError").release();
  DtExc_OverflowError         = dx.get_attr("OverflowError").release();
  DtExc_TypeError             = dx.get_attr("TypeError").release();
  DtExc_ValueError            = dx.get_attr("ValueError").release();
  DtWrn_DatatableWarning      = dx.get_attr("DatatableWarning").release();
  DtWrn_IOWarning             = dx.get_attr("IOWarning").release();
}



//------------------------------------------------------------------------------
// Helper functions
//------------------------------------------------------------------------------

static bool is_string_empty(const char* msg) noexcept {
  if (!msg) return true;
  char c;
  while ((c = *msg) != 0) {
    if (!(c == ' ' || c == '\t' || c == '\n' || c == '\r'))
      return false;
    msg++;
  }
  return true;
}


void exception_to_python(const std::exception& e) noexcept {
  wassert(dt::num_threads_in_team() == 0);
  const Error* error = dynamic_cast<const Error*>(&e);
  if (error) {
    error->to_python();
  }
  else if (!PyErr_Occurred()) {
    const char* msg = e.what();
    if (is_string_empty(msg)) {
      PyErr_SetString(PyExc_Exception, "unknown error");
    } else {
      PyErr_SetString(PyExc_Exception, msg);
    }
  }
}


/**
  * If `str` contains any backticks or backslashes, they will be
  * escaped by prepending each such character with a backslash.
  * If there are no backticks/backslahes in `str`, then a simple copy
  * of the string is returned.
  */
std::string escape_backticks(const std::string& str) {
  size_t count = 0;
  for (char c : str) {
    count += (c == '`' || c == '\\');
  }
  if (count == 0) return str;
  std::string out;
  out.reserve(str.size() + count);
  for (char c : str) {
    if (c == '`' || c == '\\') out += '\\';
    out += c;
  }
  return out;
}




//------------------------------------------------------------------------------
// Error: constructors
//------------------------------------------------------------------------------

Error::Error()
  : pycls_(nullptr),
    exc_type_(nullptr),
    exc_value_(nullptr),
    exc_traceback_(nullptr)
{
  PyErr_Fetch(&exc_type_, &exc_value_, &exc_traceback_);
  if (is_keyboard_interrupt()) {
    dt::progress::manager->set_status_cancelled();
  }
}


Error::Error(PyObject* exception_class)
  : pycls_(exception_class),
    exc_type_(nullptr),
    exc_value_(nullptr),
    exc_traceback_(nullptr) {}


Error::Error(const Error& other) {
  error_message_ << other.error_message_.str();
  pycls_ = other.pycls_;
  exc_type_      = other.exc_type_;       Py_XINCREF(exc_type_);
  exc_value_     = other.exc_value_;      Py_XINCREF(exc_value_);
  exc_traceback_ = other.exc_traceback_;  Py_XINCREF(exc_traceback_);
}

Error::Error(Error&& other) : Error(nullptr) {
  std::swap(error_message_, other.error_message_);
  std::swap(pycls_, other.pycls_);
  std::swap(exc_type_, other.exc_type_);
  std::swap(exc_value_, other.exc_value_);
  std::swap(exc_traceback_, other.exc_traceback_);
}

Error::~Error() {
  Py_XDECREF(exc_type_);      exc_type_ = nullptr;
  Py_XDECREF(exc_value_);     exc_value_ = nullptr;
  Py_XDECREF(exc_traceback_); exc_traceback_ = nullptr;
}




//------------------------------------------------------------------------------
// Error: message logging methods
//------------------------------------------------------------------------------

template <>
Error& Error::operator<<(const dt::CString& str) {
  return *this << str.to_string();
}


template <>
Error& Error::operator<<(const py::robj& o) {
  return *this << o.to_borrowed_ref();
}


template <>
Error& Error::operator<<(const py::oobj& o) {
  return *this << o.to_borrowed_ref();
}


template <>
Error& Error::operator<<(const py::ostring& o) {
  PyObject* ptr = o.to_borrowed_ref();
  Py_ssize_t size;
  const char* chardata = PyUnicode_AsUTF8AndSize(ptr, &size);
  if (chardata) {
    error_message_ << std::string(chardata, static_cast<size_t>(size));
  } else {
    error_message_ << "<unknown>";
    PyErr_Clear();
  }
  return *this;
}


template <>
Error& Error::operator<<(const PyObjectPtr& v) {
  PyObject* repr = PyObject_Repr(v);
  Py_ssize_t size;
  const char* chardata = PyUnicode_AsUTF8AndSize(repr, &size);
  if (chardata) {
    error_message_ << std::string(chardata, static_cast<size_t>(size));
  } else {
    error_message_ << "<unknown>";
    PyErr_Clear();
  }
  Py_XDECREF(repr);
  return *this;
}


template <>
Error& Error::operator<<(const PyTypeObjectPtr& v) {
  return *this << reinterpret_cast<const PyObjectPtr>(v);
}


template <>
Error& Error::operator<<(const CErrno&) {
  error_message_ << "[errno " << errno << "] " << strerror(errno);
  return *this;
}


template <>
Error& Error::operator<<(const dt::SType& stype) {
  error_message_ << dt::stype_name(stype);
  return *this;
}


template <>
Error& Error::operator<<(const dt::LType& ltype) {
  error_message_ << dt::ltype_name(ltype);
  return *this;
}


template <>
Error& Error::operator<<(const dt::Type& type) {
  error_message_ << type.to_string();
  return *this;
}


template <>
Error& Error::operator<<(const char& c) {
  uint8_t uc = static_cast<uint8_t>(c);
  if (uc < 0x20 || uc >= 0x80 || uc == '`' || uc == '\\') {
    error_message_ << '\\';
    if (c == '\n') error_message_ << 'n';
    else if (c == '\r') error_message_ << 'r';
    else if (c == '\t') error_message_ << 't';
    else if (c == '\\') error_message_ << '\\';
    else if (c == '`')  error_message_ << '`';
    else {
      uint8_t d1 = uc >> 4;
      uint8_t d2 = uc & 15;
      error_message_ << "\\x" << static_cast<char>((d1 <= 9? '0' : 'a' - 10) + d1)
                     << static_cast<char>((d2 <= 9? '0' : 'a' - 10) + d2);
    }
  } else {
    error_message_ << c;
  }
  return *this;
}




//------------------------------------------------------------------------------
// Error: misc methods
//------------------------------------------------------------------------------

void Error::to_stderr() const {
  std::cerr << to_string() << "\n";
}

std::string Error::to_string() const {
  if (pycls_) {
    return error_message_.str();
  }
  else {
    py::ostring msg = py::robj(exc_value_).to_pystring_force();
    return msg.to_string();
  }
}


// This will not work with PyError though
bool Error::matches_exception_class(Error(*exc)()) const {
  return exc().pycls_ == pycls_;
}


void Error::to_python() const noexcept {
  // The pointer returned by errstr.c_str() is valid until errstr gets out
  // of scope. By contrast, `error_message_.str().c_str()` returns a dangling pointer,
  // which usually works but sometimes doesn't...
  // See https://stackoverflow.com/questions/1374468
  try {
    if (pycls_) {
      auto errstr = to_string();
      PyErr_SetString(pycls_, errstr.c_str());
    } else {
      // This call takes away a reference to each object: you must own a
      // reference to each object before the call and after the call you no
      // longer own these references.
      PyErr_Restore(exc_type_, exc_value_, exc_traceback_);
      exc_type_ = nullptr;
      exc_value_ = nullptr;
      exc_traceback_ = nullptr;
    }
  }
  catch (const std::exception& e) {
    PyErr_SetString(PyExc_RuntimeError, e.what());
  }
}


bool Error::is_keyboard_interrupt() const noexcept {
  return exc_type_ == PyExc_KeyboardInterrupt;
}


void Error::emit_warning() const {
  auto errstr = to_string();
  // Normally, PyErr_WarnEx returns 0. However, when the `warnings` module is
  // configured in such a way that all warnings are converted into errors,
  // then PyErr_WarnEx will return -1. At that point we should throw
  // an exception too, the error message is already set in Python.
  int ret = PyErr_WarnEx(pycls_, errstr.c_str(), 1);
  if (ret) throw PyError();
}




//==============================================================================

Error AssertionError()        { return Error(PyExc_AssertionError); }
Error AttributeError()        { return Error(PyExc_AttributeError); }
Error RuntimeError()          { return Error(PyExc_RuntimeError); }
Error ImportError()           { return Error(DtExc_ImportError); }
Error IndexError()            { return Error(DtExc_IndexError); }
Error IOError()               { return Error(DtExc_IOError); }
Error KeyError()              { return Error(DtExc_KeyError); }
Error MemoryError()           { return Error(DtExc_MemoryError); }
Error NotImplError()          { return Error(DtExc_NotImplementedError); }
Error OverflowError()         { return Error(DtExc_OverflowError); }
Error TypeError()             { return Error(DtExc_TypeError); }
Error ValueError()            { return Error(DtExc_ValueError); }
Error InvalidOperationError() { return Error(DtExc_InvalidOperationError); }
Error PyError()               { return Error(); }

Error DeprecationWarning() { return Error(PyExc_FutureWarning); }
Error DatatableWarning()   { return Error(DtWrn_DatatableWarning); }
Error IOWarning()          { return Error(DtWrn_IOWarning); }




//------------------------------------------------------------------------------
// HidePythonError
//------------------------------------------------------------------------------

HidePythonError::HidePythonError() {
  if (PyErr_Occurred()) {
    PyErr_Fetch(&exc_type_, &exc_value_, &exc_traceback_);
  } else {
    exc_type_ = nullptr;
    exc_value_ = nullptr;
    exc_traceback_ = nullptr;
  }
}

HidePythonError::~HidePythonError() {
  if (exc_type_) {
    PyErr_Restore(exc_type_, exc_value_, exc_traceback_);
  }
}
